### ISSUE TITLE
Rule proposal: `no-new-array`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#960]
There is a [`no-array-constructor`](https://eslint.org/docs/rules/no-array-constructor) rule in core, but it's not what I want, it enforce `Array(1, 2)` and `new Array(1, 2)` to `[1, 2]`,

## Fail

```js
const foo = new Array(10);
```

## Pass

```js
const foo = Array.from({length: 10});
```

Rule should provide suggestions for unknown argument:

```js
const foo = new Array(bar);
```

Suggestion 1:

```js
const foo = [bar];
```

Suggestion 2:

```js
const foo = Array.from({length: bar});
```

**Comment by silverwind** [comment]
Hmm I don't like this `Array.from({length: 10})` because it's abusing a object to be an arrayLike.

**Comment by fisker** [comment]
You prefer `new Array(length)`?

**Comment by silverwind** [comment]
That's what I use but I hat everything involving `new` generally 😉

**Comment by sindresorhus** [comment]
The `Array.from` trick also has the benefit that it's the most efficient solution if you need to map over the array afterward as you can do such in one operation:

```js
Array.from({length: 10}, (x, i) => i);
```

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by el1f** [comment]
Doesn't this conflict with `new-for-builtins`?

**Comment by fisker** [comment]
I don't think it's conflict, if you write `Array(1)`, you'll get fix to `new Array(1)`, because you should not use `Array` as function(not everyone enables `no-new-array`), then `new Array(1)` is fixed to use `Array.from({length: 1})`.

**Comment by clshortfuse** [comment]
Performance is not negligible:

https://www.measurethat.net/Benchmarks/Show/24886/0/array-construction-comparison#latest_results_block

We're talking magnitudes slower than `new Array()`

I wouldn't make it autofix. Still,  `[].fill` is also magnitudes faster, so if we're rewriting `new Array(x)`, it should opt to that. 

Using `[].fill(undefined, 0, x)` is slower than `[].fill(void 0, 0, x)`, but that optimization can be done by code minifiers, but `new Array()` should just be `[]`.

**Comment by silverwind** [comment]
[Perf link](https://perf.link/#eyJpZCI6Im43ODJpMW9ueWpnIiwidGl0bGUiOiIiLCJiZWZvcmUiOiIiLCJ0ZXN0cyI6W3sibmFtZSI6IlRlc3QgQ2FzZSIsImNvZGUiOiJBcnJheSgxMCkiLCJydW5zIjpbMjMxOTAwMCwxNTQ1MDAwLDI1MTMwMDAsNDYwMjAwMCw5NzQwMDAsNTA2NDAwMCwzMzE3MDAwLDMyNTAwMCwyMTcwMDAsMTUxMTAwMCw0MDE1MDAwLDkxNjAwMCw2MjIyMDAwLDU0OTUwMDAsMTE0MDAwLDE0MjQwMDAsMjYxODAwMCwzNjMzMDAwLDExNTUwMDAsNTcxMDAwMCw1NjAzMDAwLDE5NzAwMCwzNDQ2MDAwLDI0MjcwMDAsNDc3NjAwMCwxNzQwMDAsMjQ5MTAwMCw0NDcyMDAwLDI1NzQwMDAsMjcxNDAwMCwzMzczMDAwLDExODcwMDAsMjg4NzAwMCwzMzAwMCwyNjI2MDAwLDI1MzAwMCw0OTc4MDAwLDQ1NTUwMDAsMzQwOTAwMCw0NzY2MDAwLDM1NzUwMDAsNjUzMDAwLDQ1NDkwMDAsNzI2MDAwLDE3OTAwMCw4ODEwMDAsNDAwNzAwMCw3NDAwMDAsMjQxMTAwMCwzODE5MDAwLDk0NzAwMCw5MDUwMDAsMTMwMDAwMCwxNTcyMDAwLDM5NjQwMDAsMzM2NTAwMCwxOTQ1MDAwLDMwNzAwMDAsNTMwMDAsNDY1NDAwMCwyNTUxMDAwLDY3NDAwMCwzODQwMDAwLDEwNDUwMDAsNDY4NjAwMCwxMTg0MDAwLDUxMzAwMCwzOTk5MDAwLDIwNDAwMCwyODI2MDAwLDQ5MjYwMDAsMTUzMjAwMCwyMDI3MDAwLDM3MTYwMDAsNDQxMTAwMCwzOTU0MDAwLDIwMDAwMCw4MTAwMDAsMTk0MjAwMCwzMjQ3MDAwLDM3NjgwMDAsMzI2MDAwMCwyNjc3MDAwLDMyNTAwMCwzNjg4MDAwLDUwMzkwMDAsNDQ3NDAwMCwxMDY4MDAwLDQyMjcwMDAsMTU5MjAwMCwzNjc5MDAwLDIxMDAwLDIxMDQwMDAsMTU2NTAwMCwzODA5MDAwLDI0NjEwMDAsMzIzMTAwMCwyOTc1MDAwLDMyOTAwMDAsMTMyMDAwMF0sIm9wcyI6MjU2ODA1MH0seyJuYW1lIjoiRmluZCBpdGVtIDgwMCIsImNvZGUiOiJBcnJheS5mcm9tKHtsZW5ndGg6IDEwfSk7IiwicnVucyI6Wzk3MDAwLDEwNTAwMCwzNzUwMDAsMzU0MDAwLDIzMDAwLDQwMzAwMCw3MDgwMDAsNDYzMDAwLDMzNjAwMCwyNjgwMDAsMjIwMDAwLDI3NjAwMCw2NDAwMDAsNTQ2MDAwLDQ0MTAwMCw2MTkwMDAsMzQwMDAwLDU3MTAwMCw0OTUwMDAsNTYyMDAwLDI0ODAwMCwzMzcwMDAsODAwMDAsMTEwMDAsMTQwMDAsNTgwMDAsNzk2MDAwLDI0MjAwMCw1NzAwMDAsOTQxMDAwLDY5MTAwMCw2NDUwMDAsMjA3MDAwLDUxMDAwMCw0MzcwMDAsMTIzMDAwLDM5NDAwMCw1NTMwMDAsODEwMDAsODczMDAwLDg5MzAwMCw0NTMwMDAsODQ4MDAwLDQ4NjAwMCwxMzkwMDAsMzk1MDAwLDQ5MDAwLDE3OTAwMCw1NzkwMDAsMTU1MDAwLDMwNDAwMCw4NTkwMDAsMTk2MDAwLDE3NjAwMCwxMjAwMCwxNDEwMDAsNTY1MDAwLDI2MDAwMCw3NDAwMDAsODA4MDAwLDQ5NjAwMCw0NTMwMDAsMjczMDAwLDEwMDAwMCw2MTQwMDAsNTAyMDAwLDYzNTAwMCw2MDAwMDAsODE0MDAwLDUwNTAwMCw0NzQwMDAsMjA0MDAwLDIyNzAwMCw1MDIwMDAsNDExMDAwLDY4MDAwLDE4MTAwMCwyMDAwMCw0NDMwMDAsMjM4MDAwLDE3NzAwMCw0OTcwMDAsNzUwMDAwLDI3MDAwLDc2MzAwMCw3NDMwMDAsMjg4MDAwLDE3OTAwMCwzMTgwMDAsNDkzMDAwLDY3MDAwMCw3NTAwMDAsNTY0MDAwLDQ1MDAwMCw1NjYwMDAsNzYzMDAwLDc5MTAwMCw2NDcwMDAsNjk5MDAwLDI5NzAwMF0sIm9wcyI6NDIwODIwfV0sInVwZGF0ZWQiOiIyMDIzLTA0LTI1VDExOjExOjMxLjIwMloifQ%3D%3D).

**Comment by fisker** [comment]
Lock due to age.
